_source/_posts/2021-10-27-k8s-api-server-oidc.md

```diff
@@ -0,0 +1,551 @@
+
+# Why OpenID Connect?
+
+> [OpenID Connect](https://developer.okta.com/docs/concepts/oauth-openid/) is an authentication standard built on top of OAuth 2.0. It adds an additional token called an ID token. OpenID Connect also standardizes areas that OAuth 2.0 leaves up to choices, such as scopes, endpoint discovery, and dynamic registration of clients
+
+While the default authentication mechanism of the Kubernetes API server might be enough for simple use cases, where only a handful of people manages the cluster, it does not scale for bigger organizations. It's definitely not the most secure way as well. This is because Kubernetes does not handle user management and expects this to be done outside. This is where OpenID Connect (OIDC) comes in. OIDC can manage users and groups, which works very well with the Kubernetes RBAC to provide very granular control of who can access what inside a cluster.
+
+{% img blog/k8s-api-server-oidc/kube-login-oidc.jpg alt:"K8s OIDC flow" width:"800" %}{: .center-image }
+
+Having an OIDC integration also means you can use the same OIDC provider used to do SSO in your existing infrastructure to access your Kubernetes cluster as well, like Okta or Keycloak, for example.
+
+# Using Okta as an OIDC provider to secure the API server.
+
+> Okta, Inc. is an identity and access management company providing cloud software that helps companies manage and secure user authentication into applications and for developers to build identity controls into applications, websites, web services, and devices. Okta is a certified OpenID Connect provider.
+
+Let us see how we can secure the Kubernetes API server using Okta as an OIDC provider and use RBAC to control access right from the Okta admin console. If you are using Amazon EKS, then check [this](/blog/2021/10/08/secure-access-to-aws-eks) specific tutorial for using Okta OIDC with EKS.
+
+## What You'll Need to Get Started
+
+Before you start trying this out, make sure you have access to the following.
+
+- An Okta account. You can sign up for a free account [here](https://www.okta.com/free-trial/). If you like, you can use another OIDC provider or [Dex](https://github.com/dexidp/dex), and the steps should be similar.
+- A Kubernetes cluster. I'm using [k3d](https://k3d.io/) to run a local [k3s](https://k3s.io/) cluster. You can use any Kubernetes distribution, including managed PaaS like Amazon EKS, AKS, and GKE, and so on.
+- kubectl installed on your machine.
+- Terraform installed on your machine. This is not required if you do the Okta configuration via the [Okta admin console](https://login.okta.com/) GUI.
```

Comment:
```suggestion
- Terraform installed on your machine. This is not required if you do the Okta configuration via the [Okta Admin Console](https://login.okta.com/) GUI.
```
I don't know if I'd link to the admin console, but you can if you think it's a good idea.

Reply:
First time I encountered it in an Okta blog I had to figure out how to access it so I thought it would be helpful to just point to it